### PR TITLE
Fix jacobian calculation

### PR DESCRIPTION
**Pull request by sjahr**
### Description

@marioprats and I encountered an error in the Jacobian calculation for a link that is not the tip link. In that case all joint in the chain after the reference link don't have any impact on it's motion and thus, the related Jacobian entries need to zero. This PR fixes that + adds a unittest to catch this in the future.

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/moveit/moveit2/blob/main/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/moveit/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Review by marioprats — APPROVED** [review]
LGTM, and thanks for adding to the test!

**Review by sea-bass — DISMISSED** [review]
LGTM -- so these values were formerly undefined? Good catch.

(something something use Pinocchio instead)

**Comment by sea-bass** [comment]
I really need to look at CI results before reviewing... seems like this fails on Jazzy/Rolling for some reason?

**Comment by sjahr** [comment]
> I really need to look at CI results before reviewing... seems like this fails on Jazzy/Rolling for some reason?

Wow, this is bad :see_no_evil: Why would it fail on rolling and iron but not on Humble. There is not ROS supposed to be involved in core

**Comment by sea-bass** [comment]
> Why would it fail on rolling and iron but not on Humble. There is not ROS supposed to be involved in core

No, but new Ubuntu version might mean newer dependencies like Eigen etc.

Probably best to repro these locally on a 24.04 setup

File: moveit_core/robot_state/src/robot_state.cpp [hunk 1]
```diff
@@ -1496,19 +1496,25 @@ bool RobotState::getJacobian(const JointModelGroup* group, const LinkModel* link
   const int rows = use_quaternion_representation ? 7 : 6;
   const int columns = group->getVariableCount();
```
Comment (sea-bass) [inline moveit_core/robot_state/src/robot_state.cpp:1497]: ah! columns should not be the GROUP's variable count. Should be the entire robot?

**Comment by sjahr** [comment]
Very dump mistake, let's see if this fixes it. Interestingly, this error did not seem to happen in older Eigen versions :thinking:

**Review by sea-bass — APPROVED** [review]
Oh --  it was the test! Nice.

**Review by rhaschke — COMMENTED** [review]
The new code implicitly assumes a linear kinematic chain. It does not work for a tree, like two arms or multi-fingered hands.

**Comment by sea-bass** [comment]
> The new code implicitly assumes a linear kinematic chain. It does not work for a tree, like two arms or multi-fingered hands.

Good catch! This would also be a good unit test to add.

**Comment by sea-bass** [comment]
> > The new code implicitly assumes a linear kinematic chain. It does not work for a tree, like two arms or multi-fingered hands.
> 
> Good catch! This would also be a good unit test to add.

Update: I just tried a simple test case to see what happened, and there is a specific error that tells you getting Jacobians from non-chain groups is already not supported. So I guess we should make a separate issue to support this.

https://github.com/moveit/moveit2/blob/4c28a74343e1d8470689cae9e23eb30c91bdcb4d/moveit_core/robot_state/src/robot_state.cpp#L1467-L1472

Also... Pinocchio, anyone? :)

**Pull request merged**